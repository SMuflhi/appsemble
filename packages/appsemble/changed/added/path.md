Add property `path` to `BootstrapParams`
