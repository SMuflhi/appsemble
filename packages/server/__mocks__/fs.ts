import { fs } from 'memfs';

/**
 * This mock replaces {@link fs} with the default instnce of {@link memfs}.
 *
 * Make sure to reset before or after every test in a test file.
 *
 * @example
 * import { vol } from 'memfs';
 *
 * jest.mock('fs');
 *
 * beforeEach(() => {
 *   vol.fromJson({
 *     'package.json': JSON.stringify({})
 *   });
 * });
 */
module.exports = fs;
