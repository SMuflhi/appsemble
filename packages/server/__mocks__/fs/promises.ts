import { fs } from 'memfs';

/**
 * This mock replaces {@link fs} with the default instnce of {@link memfs}.
 *
 * Make sure to reset before or after every test in a test file.
 *
 * @example
 * import { vol } from 'memfs';
 *
 * jest.mock('fs/promises');
 *
 * beforeEach(() => {
 *   vol.fromJson({
 *     'package.json': JSON.stringify({})
 *   });
 * });
 */
export = fs.promises;
