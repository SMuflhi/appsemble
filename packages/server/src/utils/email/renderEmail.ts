import yaml from 'js-yaml';
import type { InlineCode, Link, Parent } from 'mdast';
import rehypeDocument from 'rehype-document';
import rehypeStringify from 'rehype-stringify';
import frontmatter, { YamlNode } from 'remark-frontmatter';
import remarkParse from 'remark-parse';
import remarkRehype from 'remark-rehype';
import remarkStringify from 'remark-stringify';
import unified from 'unified';
import visit from 'unist-util-visit';

const remark = unified().use(remarkParse).use(frontmatter).use(remarkStringify).freeze();

const rehype = unified().use(remarkRehype).use(rehypeDocument).use(rehypeStringify).freeze();

interface Email {
  html: string;
  subject: string;
  text: string;
}

/**
 * Render a markdown email.
 *
 * @param template The body of the template to render.
 * @param values Values to pass to the template for rendering.
 * @param sub - The subject of the email to send. If omitted, this is extracted from the markdown
 * email body.
 * @returns An email object that may be sent.
 */
export default async function renderEmail(
  template: string,
  values: { [key: string]: string },
  sub?: string,
): Promise<Email> {
  let subject = sub;
  const mdast = await remark.parse(template);

  function replace(_match: string, key: string): string {
    if (Object.prototype.hasOwnProperty.call(values, key)) {
      return values[key];
    }
    throw new Error(`Unknown template value: ${key}`);
  }

  visit<Link>(mdast, 'link', (node) => {
    // eslint-disable-next-line no-param-reassign
    node.url = node.url.replace(/\/{{(\w+)}}/, replace);
  });
  visit<InlineCode>(mdast, 'inlineCode', (node, index, parent: Parent) => {
    parent.children.splice(index, 1, {
      type: 'text',
      value: node.value.replace(/{{(\w+)}}/, replace),
    });
  });

  if (!sub) {
    visit<YamlNode>(mdast, 'yaml', (node, index, parent: Parent) => {
      ({ subject } = yaml.safeLoad(node.value) as Email);
      parent.children.splice(index, 1);
    });
  }

  const text = await remark.stringify(mdast);
  const html = await rehype.stringify(await rehype.run(mdast, { stem: subject }));
  return { html, subject, text };
}
