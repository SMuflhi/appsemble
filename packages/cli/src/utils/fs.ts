import { Dirent, promises as fs, Stats } from 'fs';
import { join } from 'path';

import { AppsembleError } from '@appsemble/node-utils';
import type { Promisable } from 'type-fest';

interface OpenDirSafeOptions {
  allowMissing?: boolean;
}

/**
 * Read the contents of a directory.
 *
 * @param directory - The path of the directory to open.
 * @param onFile - A callback which will get called for every file. This will be called with the
 * full file path as its first argument and its `Dirent` object as the second argument.
 * @param options - Additional options.
 */
export async function opendirSafe(
  directory: string,
  onFile: (fullpath: string, stat: Dirent) => Promisable<void>,
  { allowMissing }: OpenDirSafeOptions = {},
): Promise<void> {
  let stats: Stats;
  try {
    stats = await fs.stat(directory);
  } catch (err: unknown) {
    if (allowMissing && (err as NodeJS.ErrnoException).code === 'ENOENT') {
      return;
    }
    throw new AppsembleError(`Expected ${directory} to be a directory`);
  }
  if (!stats.isDirectory()) {
    throw new AppsembleError(`Expected ${directory} to be a directory`);
  }
  const dir = await fs.opendir(directory);
  for await (const file of dir) {
    await onFile(join(directory, file.name), file);
  }
}
