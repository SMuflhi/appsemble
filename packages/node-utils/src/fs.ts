import { Dirent, promises as fs, Stats } from 'fs';
import { join } from 'path';

import yaml from 'js-yaml';
import { Promisable } from 'type-fest';

import { AppsembleError } from '.';

/**
 * Test if the error is a NodeJS errno exception.
 *
 * @param error - The value to check
 * @param code - If specified, theck if the code matches
 * @returns Whether or not the error is a NodeJS errno exception.
 */
export function isErrno(error: unknown, code?: string): error is NodeJS.ErrnoException {
  if (!error || typeof error !== 'object') {
    return false;
  }
  const err = error as NodeJS.ErrnoException;
  if (code) {
    return err.code === code;
  }
  return typeof err.code === 'string';
}

/**
 * Read and parse a YAML file.
 *
 * @param path - The path to the file to read.
 * @returns A tuple of the parsed YAML content and the YAML content as a string.
 */
export async function readYaml<R>(path: string): Promise<[R, string]> {
  let content: string;
  try {
    content = await fs.readFile(path, 'utf8');
  } catch {
    throw new AppsembleError(`Error reading file ${path}`);
  }
  try {
    return [(yaml.safeLoad(content) as unknown) as R, content];
  } catch (error: unknown) {
    if (error instanceof yaml.YAMLException) {
      throw new AppsembleError(`Error parsing ${path}\n${error.message}`);
    }
    throw error;
  }
}

interface OpenDirSafeOptions {
  allowMissing?: boolean;
  recursive?: boolean;
}

/**
 * Read the contents of a directory.
 *
 * @param directory - The path of the directory to open.
 * @param onFile - A callback which will get called for every file. This will be called with the
 * full file path as its first argument and its `Dirent` object as the second argument.
 * @param options - Additional options.
 */
export async function opendirSafe(
  directory: string,
  onFile: (fullpath: string, stat: Dirent) => Promisable<void>,
  options: OpenDirSafeOptions = {},
): Promise<void> {
  let stats: Stats;
  try {
    stats = await fs.stat(directory);
  } catch (err: unknown) {
    if (options.allowMissing && isErrno(err, 'ENOENT')) {
      return;
    }
    throw new AppsembleError(`Expected ${directory} to be a directory`);
  }
  if (!stats.isDirectory()) {
    throw new AppsembleError(`Expected ${directory} to be a directory`);
  }
  const dir = await fs.opendir(directory);
  for await (const file of dir) {
    const fullPath = join(directory, file.name);
    await onFile(fullPath, file);
    if (options.recursive && file.isDirectory()) {
      await opendirSafe(fullPath, onFile, options);
    }
  }
}
