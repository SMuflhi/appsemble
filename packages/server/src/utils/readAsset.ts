import { promises as fs } from 'fs';
import path from 'path';

export const assetDir = path.resolve(__dirname, '../../assets');

/**
 * Read a file from the server assets directory.
 *
 * @param filename - The path to the file to read, relative to the assets directory.
 * @param encoding - If specified, decode the file using this encoding.
 *
 * @returns The content of the specified file.
 */
export function readAsset(filename: string, encoding?: 'utf-8'): Promise<string | Buffer> {
  return fs.readFile(path.join(assetDir, filename), encoding);
}
