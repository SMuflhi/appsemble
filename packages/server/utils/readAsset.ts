import { readFile } from 'fs/promises';
import { join, resolve } from 'path';

export const assetDir = resolve(__dirname, '../assets');

/**
 * Read a file from the server assets directory.
 *
 * @param filename The path to the file to read, relative to the assets directory.
 * @param encoding If specified, decode the file using this encoding.
 * @returns The content of the specified file.
 */
export function readAsset(filename: string): Promise<Buffer>;
export function readAsset(filename: string, encoding: BufferEncoding): Promise<string>;
export function readAsset(filename: string, encoding?: BufferEncoding): Promise<Buffer | string> {
  return readFile(join(assetDir, filename), encoding);
}
