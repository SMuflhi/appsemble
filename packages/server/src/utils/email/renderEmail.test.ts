import fs from 'fs';
import path from 'path';

import { assetDir } from '../readAsset';
import renderEmail from './renderEmail';

const tests = {
  emailAdded: [
    { greeting: 'Hello User', url: 'https://example.com' },
    { greeting: 'Hello', url: 'https://example.com' },
  ],
  organizationInvite: [
    { greeting: 'Hello User', url: 'https://example.com', organization: 'winners' },
    { greeting: 'Hello', url: 'https://example.com', organization: 'winners' },
  ],
  resend: [
    { greeting: 'Hello User', url: 'https://example.com' },
    { greeting: 'Hello', url: 'https://example.com' },
  ],
  reset: [
    { greeting: 'Hello User', url: 'https://example.com' },
    { greeting: 'Hello', url: 'https://example.com' },
  ],
  welcome: [
    { greeting: 'Hello User', url: 'https://example.com' },
    { greeting: 'Hello', url: 'https://example.com' },
  ],
};

it.each(fs.readdirSync(path.join(assetDir, 'email')).map((f) => path.parse(f).name))(
  'should have tests for %s',
  (name) => {
    expect(tests).toHaveProperty(name);
    expect(tests[name as keyof typeof tests].length).toBeGreaterThan(0);
  },
);

describe.each(Object.entries(tests))('%s', (name, testValues) => {
  it.each(testValues)(`should render ${name} %#`, async (values) => {
    const template = fs.readFileSync(path.join(assetDir, 'email', `${name}.md`), 'utf8');
    const { html, text } = await renderEmail(template, values);
    expect(text).toMatchSnapshot('text');
    expect(html).toMatchSnapshot('html');
  });
});
