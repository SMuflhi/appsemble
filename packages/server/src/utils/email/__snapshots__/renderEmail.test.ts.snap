// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emailAdded should render emailAdded %#: html 1`] = `
"<!doctype html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
</head>
<body>
<p>Hello User,</p>
<p>You have registered a new email address to your account. In order to make full use of this email, we
need to verify your email address.</p>
<p>Please click <a href=\\"https://example.com\\">this link</a> to verify your email address.</p>
<p><strong>Didn’t request this email?</strong></p>
<p>No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will be done with this email address.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`emailAdded should render emailAdded %#: html 2`] = `
"<!doctype html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
</head>
<body>
<p>Hello,</p>
<p>You have registered a new email address to your account. In order to make full use of this email, we
need to verify your email address.</p>
<p>Please click <a href=\\"https://example.com\\">this link</a> to verify your email address.</p>
<p><strong>Didn’t request this email?</strong></p>
<p>No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will be done with this email address.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`emailAdded should render emailAdded %#: text 1`] = `
"Hello User,

You have registered a new email address to your account. In order to make full use of this email, we
need to verify your email address.

Please click [this link](https://example.com) to verify your email address.

**Didn’t request this email?**

No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will be done with this email address.

Kind regards,

_The Appsemble Team_
"
`;

exports[`emailAdded should render emailAdded %#: text 2`] = `
"Hello,

You have registered a new email address to your account. In order to make full use of this email, we
need to verify your email address.

Please click [this link](https://example.com) to verify your email address.

**Didn’t request this email?**

No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will be done with this email address.

Kind regards,

_The Appsemble Team_
"
`;

exports[`organizationInvite should render organizationInvite %#: html 1`] = `
"<!doctype html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
</head>
<body>
<p>Hello User,</p>
<p>You have been invited to join the organization <strong>winners</strong>.</p>
<p>Joining an organization allows you to view private applications as well as create and edit apps
belonging to an organization.</p>
<p>You can accept or decline this invitation by following <a href=\\"https://example.com\\">this link</a>.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`organizationInvite should render organizationInvite %#: html 2`] = `
"<!doctype html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
</head>
<body>
<p>Hello,</p>
<p>You have been invited to join the organization <strong>winners</strong>.</p>
<p>Joining an organization allows you to view private applications as well as create and edit apps
belonging to an organization.</p>
<p>You can accept or decline this invitation by following <a href=\\"https://example.com\\">this link</a>.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`organizationInvite should render organizationInvite %#: text 1`] = `
"Hello User,

You have been invited to join the organization **winners**.

Joining an organization allows you to view private applications as well as create and edit apps
belonging to an organization.

You can accept or decline this invitation by following [this link](https://example.com).

Kind regards,

_The Appsemble Team_
"
`;

exports[`organizationInvite should render organizationInvite %#: text 2`] = `
"Hello,

You have been invited to join the organization **winners**.

Joining an organization allows you to view private applications as well as create and edit apps
belonging to an organization.

You can accept or decline this invitation by following [this link](https://example.com).

Kind regards,

_The Appsemble Team_
"
`;

exports[`resend should render resend %#: html 1`] = `
"<!doctype html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
</head>
<body>
<p>Hello User,</p>
<p>Someone has requested to resend the verification key for your account. Before you can use your
account, we need to verify your email address.</p>
<p>Please click <a href=\\"https://example.com\\">this link</a> below to verify your email address</p>
<p><strong>Didn’t request this email?</strong></p>
<p>No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will happen to your account.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`resend should render resend %#: html 2`] = `
"<!doctype html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
</head>
<body>
<p>Hello,</p>
<p>Someone has requested to resend the verification key for your account. Before you can use your
account, we need to verify your email address.</p>
<p>Please click <a href=\\"https://example.com\\">this link</a> below to verify your email address</p>
<p><strong>Didn’t request this email?</strong></p>
<p>No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will happen to your account.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`resend should render resend %#: text 1`] = `
"Hello User,

Someone has requested to resend the verification key for your account. Before you can use your
account, we need to verify your email address.

Please click [this link](https://example.com) below to verify your email address

**Didn’t request this email?**

No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will happen to your account.

Kind regards,

_The Appsemble Team_
"
`;

exports[`resend should render resend %#: text 2`] = `
"Hello,

Someone has requested to resend the verification key for your account. Before you can use your
account, we need to verify your email address.

Please click [this link](https://example.com) below to verify your email address

**Didn’t request this email?**

No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will happen to your account.

Kind regards,

_The Appsemble Team_
"
`;

exports[`reset should render reset %#: html 1`] = `
"<!doctype html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
</head>
<body>
<p>Hello User,</p>
<p>A password reset has been requested for you account. Please click <a href=\\"https://example.com\\">here</a> to reset your
password.</p>
<p><strong>Didn’t request this email?</strong></p>
<p>Your email address may have been entered by mistake. This email can be ignored safely.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`reset should render reset %#: html 2`] = `
"<!doctype html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
</head>
<body>
<p>Hello,</p>
<p>A password reset has been requested for you account. Please click <a href=\\"https://example.com\\">here</a> to reset your
password.</p>
<p><strong>Didn’t request this email?</strong></p>
<p>Your email address may have been entered by mistake. This email can be ignored safely.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`reset should render reset %#: text 1`] = `
"Hello User,

A password reset has been requested for you account. Please click [here](https://example.com) to reset your
password.

**Didn’t request this email?**

Your email address may have been entered by mistake. This email can be ignored safely.

Kind regards,

_The Appsemble Team_
"
`;

exports[`reset should render reset %#: text 2`] = `
"Hello,

A password reset has been requested for you account. Please click [here](https://example.com) to reset your
password.

**Didn’t request this email?**

Your email address may have been entered by mistake. This email can be ignored safely.

Kind regards,

_The Appsemble Team_
"
`;

exports[`welcome should render welcome %#: html 1`] = `
"<!doctype html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
</head>
<body>
<p>Hello User,</p>
<p>Thank you for registering your account. Before you can use your account, we need to verify your
email address.</p>
<p>Please click <a href=\\"https://example.com\\">here</a> to verify your email address.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`welcome should render welcome %#: html 2`] = `
"<!doctype html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
</head>
<body>
<p>Hello,</p>
<p>Thank you for registering your account. Before you can use your account, we need to verify your
email address.</p>
<p>Please click <a href=\\"https://example.com\\">here</a> to verify your email address.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`welcome should render welcome %#: text 1`] = `
"Hello User,

Thank you for registering your account. Before you can use your account, we need to verify your
email address.

Please click [here](https://example.com) to verify your email address.

Kind regards,

_The Appsemble Team_
"
`;

exports[`welcome should render welcome %#: text 2`] = `
"Hello,

Thank you for registering your account. Before you can use your account, we need to verify your
email address.

Please click [here](https://example.com) to verify your email address.

Kind regards,

_The Appsemble Team_
"
`;
